pkg/storageops/vsphere/vsphere.go

```diff
@@ -0,0 +1,397 @@
+	if !ok {
+		return "", fmt.Errorf("invalid input: %v to GetDeviceID", diskPath)
+	}
+
+	return id, nil
+}
+
+// Attach takes in the path of the vmdk file and returns where it is attached inside the vm instance
+func (ops *vsphereOps) Attach(diskPath string) (string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return "", err
+	}
+
+	diskUUID, err := ops.vm.AttachDisk(ctx, diskPath, &vclib.VolumeOptions{SCSIControllerType: vclib.PVSCSIControllerType})
+	if err != nil {
+		logrus.Errorf("Failed to attach vsphere disk: %s for VM: %s. err: +%v", diskPath, ops.vm.Name(), err)
+		return "", err
+	}
+
+	return path.Join(diskByIDPath, diskSCSIPrefix+diskUUID), nil
+}
+
+func (ops *vsphereOps) Detach(diskPath string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return err
+	}
+
+	err = ops.vm.DetachDisk(ctx, diskPath)
+	if err != nil {
+		logrus.Errorf("Failed to detach vsphere disk: %s for VM: %s. err: +%v", diskPath, ops.vm.Name(), err)
+		return err
+	}
+
+	return nil
+}
+
+// Delete virtual disk at given path
+func (ops *vsphereOps) Delete(diskPath string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return err
+	}
+
+	disk := diskmanagers.VirtualDisk{
+		DiskPath:      diskPath,
+		VolumeOptions: &vclib.VolumeOptions{},
+		VMOptions:     &vclib.VMOptions{},
+	}
+	err = disk.Delete(ctx, ops.vm.Datacenter)
+	if err != nil {
+		logrus.Errorf("Failed to delete vsphere disk: %s. err: %+v", diskPath, err)
+	}
+
+	return err
+}
+
+// Desribe an instance
+func (ops *vsphereOps) Describe() (interface{}, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return nil, err
+	}
+
+	return ops.vm, nil
+}
+
+// FreeDevices is not supported by this provider
+func (ops *vsphereOps) FreeDevices(blockDeviceMappings []interface{}, rootDeviceName string) ([]string, error) {
+	return nil, storageops.ErrNotSupported
+}
+
+func (ops *vsphereOps) Inspect(diskPaths []*string) ([]interface{}, error) {
+	// TODO find a way to map diskPaths to unattached/attached virtual disks and query info
+	// currently returning the disks directly
+
+	return nil, storageops.ErrNotSupported
+}
+
+// DeviceMappings returns map[local_attached_volume_path]->volume ID/NAME
+func (ops *vsphereOps) DeviceMappings() (map[string]string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return nil, err
+	}
+
+	vmDevices, err := ops.vm.Device(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get devices for vm: %s", ops.vm.Name())
+	}
+
+	dsMatcher := regexp.MustCompile(fmt.Sprintf(".*\\[%s.+", ops.cfg.Datastore))
```

Comment:
nit: make this regex a part of ops itself so that we do not compile it every time.